core/src/validator.rs

```diff
@@ -2181,6 +2183,63 @@ fn maybe_warp_slot(
     Ok(())
 }
 
+/// Returns the starting slot at which the blockstore should be scanned for
+/// shreds with an incorrect shred version, or None if the check is unnecessary
+fn should_check_blockstore_for_incorrect_shred_version(
+    config: &ValidatorConfig,
+    blockstore: &Blockstore,
+    root_slot: Slot,
+    hard_forks: &HardForks,
+) -> Result<Option<Slot>, BlockstoreError> {
+    // Perform the check if we are booting as part of a cluster restart at slot root_slot
+    let maybe_cluster_restart_slot = maybe_cluster_restart_with_hard_fork(config, root_slot);
+    if maybe_cluster_restart_slot.is_some() {
+        return Ok(Some(root_slot + 1));
+    }
+
+    // If there are no hard forks, the shred version cannot have changed
+    let Some(latest_hard_fork) = hard_forks.iter().last().map(|(slot, _)| *slot) else {
+        return Ok(None);
+    };
+
+    // If the blockstore is empty, there are certainly no shreds with an incorrect version
+    let Some(blockstore_max_slot) = blockstore.highest_slot()? else {
+        return Ok(None);
+    };
+    let blockstore_min_slot = blockstore.lowest_slot();
+
+    if latest_hard_fork < blockstore_min_slot {
+        // latest_hard_fork < blockstore_min_slot <= blockstore_max_slot
+        //
+        // All slots in the blockstore are newer than the latest hard fork, and only shreds with
+        // the correct shred version should have been inserted since the latest hard fork
+        //
+        // This is the normal case where the last cluster restart & hard fork was a while ago; we
+        // can skip the check for this case
+        Ok(None)
```

Comment:
For this case, can you remind how the ledger cleanup service works? I'm thinking of a scenario where it has taken 8+ hours for us to diagnose and prepare a patch with a restart slot.

Is it possible that during this time a node that did not go offline could have cleaned up everything prior to the last OC slot? This node won't be making any new roots so not sure if we still cleanup in this case.

Reply:
> This node won't be making any new roots so not sure if we still cleanup in this case.

Correct, no cleanup if roots are not advancing; we bail out of the function early as you can see below:

https://github.com/anza-xyz/agave/blob/1334fb5248390dfdd193feeec5fa8f86763668fc/ledger/src/blockstore_cleanup_service.rs#L204-L208